kong/clustering.lua

```diff
@@ -0,0 +1,424 @@
+  if not wb then
+    ngx_log(ngx_ERR, "failed to perform server side WebSocket handshake: ", err)
+    return ngx_exit(444)
+  end
+
+  local sem = semaphore.new()
+  local queue = { sem = sem, }
+  clients[wb] = queue
+
+  local res
+  -- unconditionally send config update to new clients to
+  -- ensure they have latest version running
+  res, err = declarative.export_config()
+  if not res then
+    ngx_log(ngx_ERR, "unable to export config from database: ".. err)
+  end
+
+  table.insert(queue, res)
+  queue.sem:post()
+
+  -- connection established
+  -- ping thread
+  ngx.thread.spawn(function()
+    while true do
+      local data, typ, err = wb:recv_frame()
+      if not data then
+        ngx_log(ngx_ERR, "did not receive ping frame from data plane: ", err)
+        -- return and let the main thread handle the error
+        return
+      end
+
+      assert(typ == "ping")
+      local _
+      _, err = wb:send_pong()
+      if err then
+        ngx_log(ngx_ERR, "failed to send PONG back to data plane: ", err)
+        -- return and let the main thread handle the error
+        return
+      end
+
+      ngx_log(ngx_DEBUG, "sent PONG packet to control plane")
+
+      local ok
+      ok, err = shdict:safe_set(node_id,
+                                cjson_encode({
+                                  last_seen = ngx_time(),
+                                  config_hash =
+                                    data ~= "" and data or nil,
+                                  hostname = node_hostname,
+                                  ip = node_ip,
+                                }), PING_INTERVAL * 2 + 5)
+      if not ok then
+        ngx_log(ngx_ERR, "unable to update in-memory cluster status: ", err)
+      end
+    end
+  end)
+
+  while not exiting() do
+    local ok, err = sem:wait(10)
+    if ok then
+      local config = table.remove(queue, 1)
+      assert(config, "config queue can not be empty after semaphore returns")
+
+      local _, err = wb:send_binary(cjson_encode({ type = "reconfigure",
+                                                       config_table = config,
+                                                     }))
+      if err then
+        ngx_log(ngx_ERR, "unable to send updated configuration to node: ", err)
+
+      else
+        ngx_log(ngx_DEBUG, "sent config update to node")
+      end
+
+    else -- not ok
+      if err ~= "timeout" then
+        ngx_log(ngx_ERR, "semaphore wait error: ", err)
+      end
+    end
+  end
+end
+
+
+function _M.get_status()
+  local result = new_tab(0, 8)
+
+  for _, n in ipairs(shdict:get_keys()) do
+    result[n] = cjson_decode(shdict:get(n))
+  end
+
+
+  return result
+end
+
+
+local function push_config(config_table)
+  local n = 0
+
+  for _, queue in pairs(clients) do
+    table.insert(queue, config_table)
+    queue.sem:post()
+
+    n = n + 1
+  end
+
+  ngx_log(ngx_DEBUG, "config pushed to ", n, " clients")
+end
+
+
+function _M.broadcast_config(config_table)
```

Comment:
Question: Where is this function called from?

Reply:
Good catch, that was a leftover from early development, it has been removed :)